File: api/embed.py
import json
import math
import os
import re
from http.server import BaseHTTPRequestHandler
from pathlib import Path

from openai import OpenAI, OpenAIError


EMBEDDING_MODEL = "text-embedding-3-small"
EMBEDDING_DIMENSIONS = 512
PUZZLE_PATH = Path(
    os.environ.get("PUZZLE_EMBEDDINGS_PATH", "data/puzzle_embeddings_default.json")
)
FALLBACK_PUZZLE_PATH = Path("data/puzzle_embeddings_panache.json")
CORS_ALLOW_ORIGIN = os.environ.get("CORS_ALLOW_ORIGIN", "*")

# Ephemeral per-instance cache (Vercel instances are stateless).
EMBEDDING_CACHE = {}


def normalize_word(value):
    return re.sub(r"[^a-z-]", "", value.strip().lower())


def embedding_text(word):
    return f"Meaning and associations of the word: {word}"


def cosine_similarity(a, b):
    dot = sum(x * y for x, y in zip(a, b))
    mag_a = math.sqrt(sum(x * x for x in a))
    mag_b = math.sqrt(sum(y * y for y in b))
    return dot / (mag_a * mag_b)


def read_puzzle():
    path = PUZZLE_PATH if PUZZLE_PATH.exists() else FALLBACK_PUZZLE_PATH
    return json.loads(path.read_text(encoding="utf-8"))


def semantic_profile(embedding):
    puzzle = read_puzzle()
    embeddings = puzzle["embeddings"]
    anchor_words = [puzzle["center"]["word"]]
    anchor_words.extend(target["word"] for target in puzzle["targets"])

    return {
        word: round(cosine_similarity(embedding, embeddings[word]), 4)
        for word in anchor_words
    }


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        self.send_cors_headers()
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.send_header("Access-Control-Max-Age", "86400")
        self.end_headers()

    def do_POST(self):
        content_length = int(self.headers.get("Content-Length", 0))
        raw_body = self.rfile.read(content_length)

        try:
            body = json.loads(raw_body.decode("utf-8"))
        except json.JSONDecodeError:
            self.send_json({"error": "Invalid JSON body"}, status=400)
            return

        word = normalize_word(body.get("word", ""))
        if not word:
            self.send_json({"error": "Missing word"}, status=400)
            return

        if word in EMBEDDING_CACHE:
            embedding = EMBEDDING_CACHE[word]
            self.send_json(
                {
                    "word": word,
                    "embedding": embedding,
                    "profile": semantic_profile(embedding),
                    "cached": True,
                }
            )
            return

        if not os.environ.get("OPENAI_API_KEY"):
            self.send_json(
                {
                    "error": "Missing OPENAI_API_KEY in Vercel project environment variables."
                },
                status=500,
            )
            return

        try:
            client = OpenAI()
            response = client.embeddings.create(
                model=EMBEDDING_MODEL,
                input=embedding_text(word),
                dimensions=EMBEDDING_DIMENSIONS,
                encoding_format="float",
            )
        except OpenAIError as error:
            self.send_json({"error": str(error)}, status=500)
            return

        embedding = [round(float(value), 6) for value in response.data[0].embedding]
        EMBEDDING_CACHE[word] = embedding

        self.send_json(
            {
                "word": word,
                "embedding": embedding,
                "profile": semantic_profile(embedding),
                "cached": False,
            }
        )

    def send_json(self, payload, status=200):
        encoded = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_cors_headers()
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(encoded)))
        self.end_headers()
        self.wfile.write(encoded)

    def send_cors_headers(self):
        self.send_header("Access-Control-Allow-Origin", CORS_ALLOW_ORIGIN)
        self.send_header("Vary", "Origin")

File: api/embed.js
const fs = require("node:fs");
const path = require("node:path");

const EMBEDDING_MODEL = "text-embedding-3-small";
const EMBEDDING_DIMENSIONS = 512;
const CORS_ALLOW_ORIGIN = process.env.CORS_ALLOW_ORIGIN || "*";
const PUZZLE_EMBEDDINGS_PATH =
  process.env.PUZZLE_EMBEDDINGS_PATH || "data/puzzle_embeddings_default.json";
const FALLBACK_PUZZLE_EMBEDDINGS_PATH = "data/puzzle_embeddings_panache.json";

const embeddingCache = new Map();

function normalizeWord(value) {
  return String(value || "")
    .trim()
    .toLowerCase()
    .replace(/[^a-z-]/g, "");
}

function embeddingText(word) {
  return `Meaning and associations of the word: ${word}`;
}

function cosineSimilarity(a, b) {
  let dot = 0;
  let magA = 0;
  let magB = 0;
  const len = Math.min(a.length, b.length);

  for (let i = 0; i < len; i += 1) {
    dot += a[i] * b[i];
    magA += a[i] * a[i];
    magB += b[i] * b[i];
  }

  return dot / (Math.sqrt(magA) * Math.sqrt(magB));
}

function loadPuzzle() {
  const primary = path.join(process.cwd(), PUZZLE_EMBEDDINGS_PATH);
  const fallback = path.join(process.cwd(), FALLBACK_PUZZLE_EMBEDDINGS_PATH);
  const selected = fs.existsSync(primary) ? primary : fallback;
  return JSON.parse(fs.readFileSync(selected, "utf8"));
}

function semanticProfile(embedding) {
  const puzzle = loadPuzzle();
  const embeddings = puzzle.embeddings;
  const anchorWords = [puzzle.center.word, ...puzzle.targets.map((t) => t.word)];

  const profile = {};
  for (const word of anchorWords) {
    profile[word] = Number(cosineSimilarity(embedding, embeddings[word]).toFixed(4));
  }
  return profile;
}

function sendJson(res, status, payload) {
  res.status(status);
  res.setHeader("Access-Control-Allow-Origin", CORS_ALLOW_ORIGIN);
  res.setHeader("Vary", "Origin");
  res.setHeader("Content-Type", "application/json");
  res.send(JSON.stringify(payload));
}

function parseBody(req) {
  if (typeof req.body === "object" && req.body !== null) {
    return req.body;
  }
  if (typeof req.body === "string") {
    try {
      return JSON.parse(req.body);
    } catch {
      return null;
    }
  }
  return null;
}

module.exports = async function handler(req, res) {
  if (req.method === "OPTIONS") {
    res.status(204);
    res.setHeader("Access-Control-Allow-Origin", CORS_ALLOW_ORIGIN);
    res.setHeader("Vary", "Origin");
    res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type");
    res.setHeader("Access-Control-Max-Age", "86400");
    res.end();
    return;
  }

  if (req.method !== "POST") {
    sendJson(res, 405, { error: "Method not allowed. Use POST." });
    return;
  }

  const body = parseBody(req);
  if (!body) {
    sendJson(res, 400, { error: "Invalid JSON body" });
    return;
  }

  const word = normalizeWord(body.word);
  if (!word) {
    sendJson(res, 400, { error: "Missing word" });
    return;
  }

  if (embeddingCache.has(word)) {
    const embedding = embeddingCache.get(word);
    sendJson(res, 200, {
      word,
      embedding,
      profile: semanticProfile(embedding),
      cached: true,
    });
    return;
  }

  if (!process.env.OPENAI_API_KEY) {
    sendJson(res, 500, {
      error: "Missing OPENAI_API_KEY in Vercel project environment variables.",
    });
    return;
  }

  try {
    const response = await fetch("https://api.openai.com/v1/embeddings", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${process.env.OPENAI_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: EMBEDDING_MODEL,
        input: embeddingText(word),
        dimensions: EMBEDDING_DIMENSIONS,
        encoding_format: "float",
      }),
    });

    const data = await response.json();
    if (!response.ok) {
      sendJson(res, 500, {
        error: data.error?.message || "OpenAI embeddings request failed.",
      });
      return;
    }

    const embedding = data.data[0].embedding.map((v) => Number(Number(v).toFixed(6)));
    embeddingCache.set(word, embedding);

    sendJson(res, 200, {
      word,
      embedding,
      profile: semanticProfile(embedding),
      cached: false,
    });
  } catch (error) {
    sendJson(res, 500, { error: String(error.message || error) });
  }
};
